### DISCUSSION TITLE
Automatic package-based invalidation

### DISCUSSION THREAD

**Discussion by joelnitta** [ropensci/targets#743] [Help]
Apologies if this has already been answered somewhere... I looked but couldn't find anything.

[Chapter 8.4](https://books.ropensci.org/targets/packages.html#package-based-invalidation) of the manual describes how to set up package-based invalidation, but seems rather verbose because you need to type each package twice (`packages` and `imports`), and you have to manually specify each package. Is there a way to do this automatically for all packages? So something like `tar_option_set(track_all_pkgs = TRUE)`?

FYI my particular use-case is coming back to a project after a while, and realizing I need to update a few packages. I want all of my code to be reproducible, so I want any targets that depend on those packages to be invalidated... but I had no way of knowing earlier which packages I would need to update, so I didn't set that up in the targets options. Now I find myself needing to run everything over again from scratch because I'm not sure exactly which targets have been affected by the updates.

Follow-up: I thought of another reason I would like this feature: I often use the `package::func()` syntax in my functions, rather than loading all packages with `library()`. So it is difficult to search through my code, identify all packages loaded, and specify them all in `tar_option_set()`.

**Comment by wlandau** [comment]
I do not have plans to combine the ‘packages’ and ‘imports’ options, but declaring a set of common packages is straightforward.

common_packages <-c(…)
tar_option_set(packages = common_packages, imports = common_packages)

And if you use package::fn() syntax everywhere, then you can just set ‘imports’ without setting ‘packages’, which should also make initialization faster.

But I generally discourage more than 1 or 2 important ‘imports’ that you built yourself because the criteria for target invalidation become less clear. Packages have extensive internal function networks, S3 classes, and C++ code, and it requires digging into the internals of every package to confirm that the proper updates happen. Not every package can be tracked this way because some have cycles in the dependency graphs, and the pipeline becomes more vulnerable to false-positive invalidations.

I understand your frustration, but I also think that after a typical project is created, updates to the renv library are seldom necessary anymore, if at all. And this is one of those edge cases that could get a lot messier if we try to engineer special accommodations.

**Reply by joelnitta** [reply]
I see, thanks for the explanation.

> updates to the renv library are seldom necessary anymore, if at all

I don't find that to be the case—in my experience, updating / installing packages throughout project development is not a rare occurrence, even common in some projects—but it's good to understand your point of view with regards to the design of targets. So basically I should consider tracking packages out of the scope of targets, with some exceptions.

Also I realized I could have perhaps been a bit smarter when I updated packages with renv: before it asks if I want to apply the changes with `snapshot()`, I could have added those packages to the list of imports.

**Reply by wlandau** [reply]
One workaround for you could be a wrapper that invalidates a custom function in response to changes in the versions of explicit package dependencies. I do not plan to officially support it in `targets` because it becomes a deep rabbit hole once you consider the upstream package dependency graph, how to hash a package properly, and distributed parallel workers, but you are welcome to use it in your own code.

``` r
tar_function <- function(fun, packages = tar_function_packages(fun)) {
  packages <- sort(packages)
  versions <- purrr::map_chr(packages, ~as.character(packageVersion(.x)))
  formals(fun) <- c(
    formals(fun),
    .targets_packages = packages,
    .targets_versions = versions
  )
  fun
}

tar_function_packages <- function(fun) {
  tmp <- tempfile(fileext = ".R")
  writeLines(deparse(fun), tmp)
  renv::dependencies(path = tmp, progress = FALSE)$Package
}

# Demo:
fun <- function(files) {
  library(dplyr)
  n <- length(files)
  if (identical(n, 0L)) {
    return(null64)
  }
  hash <- digest_file64(files)
  if (identical(n, 1L)) {
    return(hash)
  }
  digest::digest(paste(hash, collapse = ""))
}

# The new function has new arguments that depend on the versions of packages installed.
tar_function(fun)
#> function (files, .targets_packages1 = "digest", .targets_packages2 = "dplyr", 
#>     .targets_versions1 = "0.6.29", .targets_versions2 = "1.0.7") 
#> {
#>     library(dplyr)
#>     n <- length(files)
#>     if (identical(n, 0L)) {
#>         return(null64)
#>     }
#>     hash <- digest_file64(files)
#>     if (identical(n, 1L)) {
#>         return(hash)
#>     }
#>     digest::digest(paste(hash, collapse = ""))
#> }
```

<sup>Created on 2022-01-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by joelnitta** [reply]
Thanks for the suggestion!

This doesn't quite fit my current use-case, as I am concerned about changes to a particular function of a loaded package affecting the custom functions that depend on it. But it does seem potentially useful, I will keep it in mind.
